### PR TITLE
[Improvement](pipeline) Cancel outdated query if original fe restarts

### PR DESCRIPTION
**Pull request by zhiqiang-hhhh**
## Further comments

If any FE restarts, queries that is emitted from this FE will be cancelled.

Implementation of https://github.com/apache/doris/issues/23704

### CODE REVIEW

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/system/Frontend.java [hunk 1]
```diff
@@ -60,6 +60,8 @@ public class Frontend implements Writable {
 
     private boolean isAlive = false;
 
+    private long startUUID = 0;
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/system/Frontend.java:63]: maybe call processUUID

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.42 seconds
 stream load tsv:          550 seconds loaded 74807831229 Bytes, about 129 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17161820999 Bytes

File: fe/fe-core/src/main/java/org/apache/doris/system/Frontend.java (hunk 1 above)
Comment (zhiqiang-hhhh) [reply fe/fe-core/src/main/java/org/apache/doris/system/Frontend.java:63]: ok

**Comment by zhiqiang-hhhh** [comment]
run CheckStyle

File: .gitignore [hunk 2]
```diff
@@ -37,6 +37,7 @@ package-lock.json
 !.idea/icon.svg
 **/.vscode/
 **/.fleet/
+compile_commands.json
```
Comment (yiguolei) [inline .gitignore:40]: why add this?

File: be/src/agent/heartbeat_server.cpp [hunk 3]
```diff
@@ -219,6 +222,69 @@ Status HeartbeatServer::_heartbeat(const TMasterInfo& master_info) {
...
+            LOG(INFO) << "Frontend " << PrintThriftNetworkAddress(dropped_fe) << " has already been dropped, "
+                      << " so going to cancel related fragments.";
+
+            abnormal_fe_infos.emplace_back((*_fe_infos.find(dropped_fe)).second._fe_info);
+            _fe_infos.erase(dropped_fe);
+        }
+
+        // Step2: notify cancel thread if any abnormal fes.
+        if (!abnormal_fe_infos.empty()) {
+            _exec_env->fragment_mgr()->add_abnormal_fe_infos_and_notify(abnormal_fe_infos);
```
Comment (yiguolei) [inline be/src/agent/heartbeat_server.cpp:284]: Not do this. It is very dangerous to do such work in heartbeat thread. For example, if there is a lock in fragment mgr, then heart beat will hang and fe will set the BE to bad. All queries and load jobs will fail.
Could add these abnormal fe infos to ExecEnv. And then in cancel worker thread, we could check the abnormal fes.

File: be/src/agent/heartbeat_server.h [hunk 4]
```diff
@@ -62,6 +63,15 @@ class HeartbeatServer : public HeartbeatServiceIf {
     TMasterInfo* _master_info;
     int64_t _fe_epoch;
 
+    ExecEnv* _exec_env;
+
+    struct FeInternalInfo {
+        TFrontendInfo _fe_info;
+        int _unknown_counter = 0;
+    };
+    // TODO(zhiqiang): make sure address is right.
```
Comment (yiguolei) [inline be/src/agent/heartbeat_server.h:72]: add  comment here  to explain network address, for example  feip:rpc_port

File: be/src/agent/heartbeat_server.cpp (hunk 3 above)
Comment (zhiqiang-hhhh) [reply be/src/agent/heartbeat_server.cpp:284]: wrong function name, there is no notify (condition variables) operation, just get a light weighted lock, assign vector to fragment_mgr, and fragment_mgr::cancel_worker will check if any query shoule be canceled, NOTE: actual cancel operation will not be performed under this lock, all threads will just do read/write,  and there is no io operation under the lock.

File: .gitignore (hunk 2 above)
Comment (zhiqiang-hhhh) [reply .gitignore:40]: i will soft link compile_commands.json from build dir to project root dir.  I will discard this change.

File: be/src/runtime/fragment_mgr.cpp [hunk 5]
```diff
@@ -1265,4 +1344,9 @@ void FragmentMgr::_setup_shared_hashtable_for_broadcast_join(
     }
 }
 
+void FragmentMgr::add_abnormal_fe_infos_and_notify(const std::vector<TFrontendInfo>& abnornal_fe_infos) {
```
Comment (yiguolei) [inline be/src/runtime/fragment_mgr.cpp:1347]: rename to  notfiy_abnormal_fes

File: be/src/runtime/exec_env.cpp [hunk 6]
```diff
@@ -28,4 +34,95 @@ ExecEnv::~ExecEnv() {}
...
+    for (const auto& dropped_fe : dropped_fes) {
+        LOG(INFO) << "Frontend " << PrintThriftNetworkAddress(dropped_fe) << " has already been dropped, remove it";
+        _frontends.erase(dropped_fe);
+    }
+}
+
+std::map<TNetworkAddress, FrontendInfo> ExecEnv::get_running_frontends() {
+    std::lock_guard<std::mutex> lg(_frontends_lock);
+    std::map<TNetworkAddress, FrontendInfo> res;
+    const int expired_duration = 10 * 1000; // 10 seconds.
```
Comment (yiguolei) [inline be/src/runtime/exec_env.cpp:95]: 10s  need to be a config in config.h. And 10s is too small, change it to 1min

**Review by yiguolei — DISMISSED** [review]
LGTM

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.99 seconds
 stream load tsv:          549 seconds loaded 74807831229 Bytes, about 129 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17162120632 Bytes

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 49.13 seconds
 stream load tsv:          536 seconds loaded 74807831229 Bytes, about 133 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162049244 Bytes

**Review by yiguolei — DISMISSED** [review]
LGTM

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 49.48 seconds
 stream load tsv:          533 seconds loaded 74807831229 Bytes, about 133 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.9 seconds inserted 10000000 Rows, about 346K ops/s
 storage size: 17162102199 Bytes

**Review by yiguolei — APPROVED** [review]
LGTM

**Review by xinyiZzz — APPROVED** [review]
LGTM

**Comment by xiaokang** [comment]
not merge this feature in 2.0

**Pull request merged**